core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala

```diff
@@ -59,20 +60,26 @@ trait DynamicReplacableConstant extends Expression {
   override final def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     // change the isNull and primitive to consts, to inline them
     val value = this.value
+
     assert(value != null || nullable, "Expected nullable as true when value is null")
     val addMutableState = !ctx.references.exists(_.asInstanceOf[AnyRef] eq this)
     if (addMutableState) {
-      literalValueRef = ctx.addReferenceObj("literal", this,
+      val literalValueRef = ctx.addReferenceObj("literal", this,
         classOf[DynamicReplacableConstant].getName)
-      isNull = ctx.freshName("isNull")
-      valueTerm = ctx.freshName("value")
+      val isNull = ctx.freshName("isNull")
+      val valueTerm = ctx.freshName("value")
+      termMap.put(ctx, TermValues(literalValueRef, isNull, valueTerm))
```

Comment:
For this case (the common one) don't need a get again below. Store the TermValues as result of the if (addMutableState) block. See the next comment below about the assertion so that result can be stored and the gets below should be removed.

Reply:
ok. Will change it.